curated_transformers/layers/attention.py

```diff
@@ -968,6 +949,87 @@ def _query_key_value(self, input: Tensor) -> Tuple[Tensor, Tensor, Tensor]:
 
         return query, key, value
 
+    def _query_key_value_separate(self, input: Tensor) -> Tuple[Tensor, Tensor, Tensor]:
+        """
+        Compute query, key and value representations for the input, using
+        separate query, key, and value projections.
+
+        :param input:
+            Input
+
+            *Shape:* ``(batch_size, seq_len, hidden_width)``
+        :returns:
+            Query, key, value
+
+            *Shape:* ``(batch_size, n_head, seq_len, width_per_head)``
+        """
+        query = self.query(input)
+        key = self.key(input)
+        value = self.value(input)
+
+        query = split_heads(query, self.attention_heads._n_query_heads)
+        key = split_heads(key, self.attention_heads._n_key_value_heads)
+        value = split_heads(value, self.attention_heads._n_key_value_heads)
+
+        return query, key, value
+
+    def _query_key_value_merged_split_before(
+        self, input: Tensor
+    ) -> Tuple[Tensor, Tensor, Tensor]:
+        """
+        Compute query, key and value representations for the input, using
+        merged query, key, and value projections. Split heads before
+        splitting query, key, and value.
+
+        :param input:
+            Input
+
+            *Shape:* ``(batch_size, seq_len, hidden_width)``
+        :returns:
+            Query, key, value
+
+            *Shape:* ``(batch_size, n_head, seq_len, width_per_head)``
+        """
+        proj = self.input(input)
+        # Same number of heads for query, key and value
+        # since we cannot share heads in this mode.
```

Comment:
Shall we add an assertion to this comment?

Reply:
This is already verified in the constructor:

https://github.com/explosion/curated-transformers/blob/3b465f9ff6815d6f9f18e932a60a52ef567e06ab/curated_transformers/layers/attention.py#L601